### ISSUE TITLE
PluginLoader throws NPE

### ISSUE THREAD

**Issue by twogee** [spotbugs/spotbugs#663]
For instance, when Ant task is invoked with classpathref and pluginlistref, [PluginLoader](https://github.com/spotbugs/spotbugs/blob/e120cb9ff1041948e0d970c0642c499aa6dba7d2/spotbugs/src/main/java/edu/umd/cs/findbugs/PluginLoader.java#L1183) throws an NPE.

**Comment by KengoTODA** [comment]
please provide enough information to reproduce this issue.

**Comment by twogee** [comment]
Here's the complete test case (presuming a project build using Maven SDL)

```
<project default="spotbugs" xmlns:ivy="antlib:org.apache.ivy.ant">
  <property name="java.dir" value="src/main/java"/>
  <property name="build.dir" value="target"/>
  <property name="build.classes" value="${build.dir}/classes"/>
  <property name="build.lib" value="${build.dir}/lib"/>
  <property name="spotbugs.version" value="3.1.4"/>
  <property name="spotbugs.reportdir" location="${build.dir}/reports/spotbugs"/>

  <target name="init-ivy">
    <property name="ivy.version" value="2.4.0"/>
    <property name="ivy.jar.url" value="http://repo1.maven.org/maven2/org/apache/ivy/ivy/${ivy.version}/ivy-${ivy.version}.jar"/>
    <property name="ivy.jar.dir" value="${build.dir}/ivy"/>
    <property name="ivy.jar.file" value="${ivy.jar.dir}/ivy.jar"/>
    <mkdir dir="${ivy.jar.dir}"/>
    <get src="${ivy.jar.url}" dest="${ivy.jar.file}" usetimestamp="true"/>
    <path id="ivy.lib.path">
      <fileset dir="${ivy.jar.dir}" includes="*.jar"/>
    </path>
    <taskdef resource="org/apache/ivy/ant/antlib.xml" uri="antlib:org.apache.ivy.ant" classpathref="ivy.lib.path"/>
  </target>

  <target name="spotbugs" depends="init-ivy">
    <path id="libs">
      <fileset dir="${build.lib}" includes="*.jar"/>
    </path>

    <ivy:cachepath organisation="com.github.spotbugs" module="spotbugs-ant" revision="${spotbugs.version}"
                   inline="true" conf="default" pathid="spotbugs.classpath"/>

    <taskdef resource="edu/umd/cs/findbugs/anttask/tasks.properties"
             classpathref="spotbugs.classpath"/>

    <mkdir dir="${spotbugs.reportdir}"/>
    <spotbugs classpathref="spotbugs.classpath"
	      pluginlistref="spotbugs.classpath"
              output="xml:withMessages"
              outputFile="${spotbugs.reportdir}/spotbugs.xml">
      <class location="${build.classes}"/>
      <sourcePath path="${java.dir}"/>
      <auxClasspath refid="libs"/>
    </spotbugs>
  </target>
</project>
```

**Comment by KengoTODA** [comment]
Thanks. Confirmed that it's reproducible with `3.1.0-RC1` ~ `v3.1.5`.

**Comment by KengoTODA** [comment]
debug log:

```
[spotbugs] Executing SpotBugs FindBugsTask from ant task
[spotbugs] Running SpotBugs...
[spotbugs] Executing '~/oracle-jdk-1.8.0_121/jre/bin/java' with arguments:
[spotbugs] '-Dfindbugs.hostApp=FBAntTask'
[spotbugs] '-classpath'
[spotbugs] '~/.ivy2/cache/com.github.spotbugs/spotbugs-ant/jars/spotbugs-ant-3.1.4.jar:~/.ivy2/cache/com.github.spotbugs/spotbugs/jars/spotbugs-3.1.4.jar:~/.ivy2/cache/org.ow2.asm/asm/jars/asm-6.2.jar:~/.ivy2/cache/org.ow2.asm/asm-analysis/jars/asm-analysis-6.2.jar:~/.ivy2/cache/org.ow2.asm/asm-tree/jars/asm-tree-6.2.jar:~/.ivy2/cache/org.ow2.asm/asm-commons/jars/asm-commons-6.2.jar:~/.ivy2/cache/org.ow2.asm/asm-util/jars/asm-util-6.2.jar:~/.ivy2/cache/org.ow2.asm/asm-xml/jars/asm-xml-6.2.jar:~/.ivy2/cache/org.apache.bcel/bcel/jars/bcel-6.2.jar:~/.ivy2/cache/net.jcip/jcip-annotations/jars/jcip-annotations-1.0.jar:~/.ivy2/cache/org.dom4j/dom4j/jars/dom4j-2.1.0.jar:~/.ivy2/cache/jaxen/jaxen/bundles/jaxen-1.1.6.jar:~/.ivy2/cache/commons-lang/commons-lang/jars/commons-lang-2.6.jar:~/.ivy2/cache/com.google.code.findbugs/jFormatString/jars/jFormatString-3.0.0.jar:~/.ivy2/cache/com.github.spotbugs/spotbugs-annotations/jars/spotbugs-annotations-3.1.4.jar:~/.ivy2/cache/com.google.code.findbugs/jsr305/jars/jsr305-3.0.2.jar'
[spotbugs] 'edu.umd.cs.findbugs.FindBugs2'
[spotbugs] '-pluginList'
[spotbugs] '~/.ivy2/cache/com.github.spotbugs/spotbugs-ant/jars/spotbugs-ant-3.1.4.jar:~/.ivy2/cache/com.github.spotbugs/spotbugs/jars/spotbugs-3.1.4.jar:~/.ivy2/cache/org.ow2.asm/asm/jars/asm-6.2.jar:~/.ivy2/cache/org.ow2.asm/asm-analysis/jars/asm-analysis-6.2.jar:~/.ivy2/cache/org.ow2.asm/asm-tree/jars/asm-tree-6.2.jar:~/.ivy2/cache/org.ow2.asm/asm-commons/jars/asm-commons-6.2.jar:~/.ivy2/cache/org.ow2.asm/asm-util/jars/asm-util-6.2.jar:~/.ivy2/cache/org.ow2.asm/asm-xml/jars/asm-xml-6.2.jar:~/.ivy2/cache/org.apache.bcel/bcel/jars/bcel-6.2.jar:~/.ivy2/cache/net.jcip/jcip-annotations/jars/jcip-annotations-1.0.jar:~/.ivy2/cache/org.dom4j/dom4j/jars/dom4j-2.1.0.jar:~/.ivy2/cache/jaxen/jaxen/bundles/jaxen-1.1.6.jar:~/.ivy2/cache/commons-lang/commons-lang/jars/commons-lang-2.6.jar:~/.ivy2/cache/com.google.code.findbugs/jFormatString/jars/jFormatString-3.0.0.jar:~/.ivy2/cache/com.github.spotbugs/spotbugs-annotations/jars/spotbugs-annotations-3.1.4.jar:~/.ivy2/cache/com.google.code.findbugs/jsr305/jars/jsr305-3.0.2.jar'
[spotbugs] '-sortByClass'
[spotbugs] '-timestampNow'
[spotbugs] '-xml:withMessages'
[spotbugs] '-sourcepath'
[spotbugs] '~/Desktop/src/main/java'
[spotbugs] '-outputFile'
[spotbugs] '~/Desktop/target/reports/spotbugs/spotbugs.xml'
[spotbugs] '-exitcode'
[spotbugs] '~/Desktop/target/classes'
[spotbugs] 
[spotbugs] The ' characters around the executable and arguments are
[spotbugs] not part of the command.
[spotbugs] Exception in thread "main" java.lang.NullPointerException
[spotbugs] 	at edu.umd.cs.findbugs.PluginLoader.getPluginDescriptor(PluginLoader.java:1183)
[spotbugs] 	at edu.umd.cs.findbugs.PluginLoader.init(PluginLoader.java:712)
[spotbugs] 	at edu.umd.cs.findbugs.PluginLoader.<init>(PluginLoader.java:207)
[spotbugs] 	at edu.umd.cs.findbugs.PluginLoader.getPluginLoader(PluginLoader.java:1367)
[spotbugs] 	at edu.umd.cs.findbugs.Plugin.addCustomPlugin(Plugin.java:667)
[spotbugs] 	at edu.umd.cs.findbugs.Plugin.addCustomPlugin(Plugin.java:659)
[spotbugs] 	at edu.umd.cs.findbugs.Plugin.loadCustomPlugin(Plugin.java:651)
[spotbugs] 	at edu.umd.cs.findbugs.Plugin.loadCustomPlugin(Plugin.java:644)
[spotbugs] 	at edu.umd.cs.findbugs.FindBugsCommandLine.handleOptionWithArgument(FindBugsCommandLine.java:148)
[spotbugs] 	at edu.umd.cs.findbugs.TextUICommandLine.handleOptionWithArgument(TextUICommandLine.java:560)
[spotbugs] 	at edu.umd.cs.findbugs.config.CommandLine.parse(CommandLine.java:335)
[spotbugs] 	at edu.umd.cs.findbugs.config.CommandLine.parse(CommandLine.java:299)
[spotbugs] 	at edu.umd.cs.findbugs.FindBugs.processCommandLine(FindBugs.java:351)
[spotbugs] 	at edu.umd.cs.findbugs.FindBugs2.main(FindBugs2.java:1175)
[spotbugs] Java Result: 1
```

**Comment by KengoTODA** [comment]
NPE at https://github.com/spotbugs/spotbugs/blob/3.1.5/spotbugs/src/main/java/edu/umd/cs/findbugs/PluginLoader.java#L1183

**Comment by KengoTODA** [comment]
This is because SpotBugs tried to throw `PluginDoesntContainMetadataException` but `classloader` was null so it throws NPE. If we have no NPE, `PluginDoesntContainMetadataException` like below will be thrown:

```
 [spotbugs] Caused by: edu.umd.cs.findbugs.PluginDoesntContainMetadataException: 
Plugin spotbugs-ant-3.1.6-SNAPSHOT.jar doesn't contain findbugs.xml; 
got jar:file:/home/kengo/.ivy2/cache/com.github.spotbugs/spotbugs/jars/spotbugs-3.1.6-SNAPSHOT.jar!/findbugs.xml from null
```

So problem is that, your plugin list contains not only `spotbugs.jar` but also `spotbugs-ant.jar`. As a workaround, please remove it from `pluginlist`.

**Comment by twogee** [comment]
Thanks for explanation. My workaround was to skip pluginlistref for the time being.

There is another problem, it looks like auxclasspath is added to Ant classloader even though it is only really needed in the forked JVM. Can this be avoided?

**Comment by mattnelson** [comment]
Seeing this same issue occasionally. I have a large reactor that is building in parallel. Due the randomness of the failures, my hunch is that this is a concurrency issue.

```
[INFO] >>> spotbugs-maven-plugin:3.1.6:check (default) > :spotbugs @ my-project >>>
[INFO] 
[INFO] --- spotbugs-maven-plugin:3.1.6:spotbugs (spotbugs) @ my-project ---
[INFO] Fork Value is true
     [java] Exception in thread "main" java.lang.NullPointerException
     [java] 	at edu.umd.cs.findbugs.PluginLoader.getPluginDescriptor(PluginLoader.java:1183)
     [java] 	at edu.umd.cs.findbugs.PluginLoader.init(PluginLoader.java:712)
     [java] 	at edu.umd.cs.findbugs.PluginLoader.<init>(PluginLoader.java:207)
     [java] 	at edu.umd.cs.findbugs.PluginLoader.getPluginLoader(PluginLoader.java:1367)
     [java] 	at edu.umd.cs.findbugs.Plugin.addCustomPlugin(Plugin.java:667)
     [java] 	at edu.umd.cs.findbugs.Plugin.addCustomPlugin(Plugin.java:659)
     [java] 	at edu.umd.cs.findbugs.Plugin.loadCustomPlugin(Plugin.java:651)
     [java] 	at edu.umd.cs.findbugs.Plugin.loadCustomPlugin(Plugin.java:644)
     [java] 	at edu.umd.cs.findbugs.FindBugsCommandLine.handleOptionWithArgument(FindBugsCommandLine.java:148)
     [java] 	at edu.umd.cs.findbugs.TextUICommandLine.handleOptionWithArgument(TextUICommandLine.java:560)
     [java] 	at edu.umd.cs.findbugs.config.CommandLine.parse(CommandLine.java:335)
     [java] 	at edu.umd.cs.findbugs.config.CommandLine.parse(CommandLine.java:299)
     [java] 	at edu.umd.cs.findbugs.FindBugs.processCommandLine(FindBugs.java:351)
     [java] 	at edu.umd.cs.findbugs.FindBugs2.main(FindBugs2.java:1175)
```

**Comment by Vampire** [comment]
This issue is still current with SpotBugs 4.7.3.
If you give a non-plugin to the `-pluginList` option, this NPE occurs.
Current stacktrace is
```exception
Exception in thread "main" java.lang.NullPointerException
	at edu.umd.cs.findbugs.PluginLoader.getPluginDescriptor(PluginLoader.java:1194)
	at edu.umd.cs.findbugs.PluginLoader.init(PluginLoader.java:737)
	at edu.umd.cs.findbugs.PluginLoader.<init>(PluginLoader.java:210)
	at edu.umd.cs.findbugs.PluginLoader.getPluginLoader(PluginLoader.java:1378)
	at edu.umd.cs.findbugs.Plugin.addCustomPlugin(Plugin.java:671)
	at edu.umd.cs.findbugs.Plugin.addCustomPlugin(Plugin.java:663)
	at edu.umd.cs.findbugs.Plugin.loadCustomPlugin(Plugin.java:655)
	at edu.umd.cs.findbugs.Plugin.loadCustomPlugin(Plugin.java:648)
	at edu.umd.cs.findbugs.FindBugsCommandLine.handleOptionWithArgument(FindBugsCommandLine.java:147)
	at edu.umd.cs.findbugs.TextUICommandLine.handleOptionWithArgument(TextUICommandLine.java:619)
	at edu.umd.cs.findbugs.config.CommandLine.parse(CommandLine.java:333)
	at edu.umd.cs.findbugs.config.CommandLine.parse(CommandLine.java:300)
	at edu.umd.cs.findbugs.FindBugs.processCommandLine(FindBugs.java:345)
	at edu.umd.cs.findbugs.FindBugs2.main(FindBugs2.java:1221)
```

It for example appears if you configure the `sb-contrib` plugin version 7.6.0 which added a dependency to `jakarta.xml.bind:jakarta.xml.bind-api` which then is given to `-pluginList` too and causes this NPE.

While I think it is a bug in the `spotbugs-gradle-plugin` that I reported at spotbugs/spotbugs-gradle-plugin#910, SpotBugs should probably nevertheless be made `NullPointerException` safe in this case and either throw a meaningful error or handle it gracefully if jars that are no plugins are specified for the option.
